Move DiscussionCourseXBlock to new location

itsjeyd opened the pull request:
This PR moves the XBlock that implements course-wide discussions for the solutions fork to `openedx/core/lib/xblock_builtin/xblock_discussion_course`, a sibling of the `openedx/core/lib/xblock_builtin/xblock_discussion` folder that contains code for the XBlock that implements inline discussions (`DiscussionXBlock`).

It is part of the Eucalyptus rebase for the solutions fork.

cf. [YONK-380](https://openedx.atlassian.net/browse/YONK-380)

Aside from moving the code, the PR also fixes an error related to rendering the block:

* Fix `render_js() got an unexpected keyword argument 'raw'` error (cf. changes to `common/templates/discussion/_discussion_course.html`)

<!-- 

**Questions**

* [ ] ~~Change package structure to mimic other XBlocks? (Put `xblock_discussion.py`, `xblock_discussion_course.py`, etc. into subfolder of `xblock_discussion`, leave `setup.py` at top level)~~

  Update: No, see https://github.com/edx-solutions/edx-platform/pull/741#issuecomment-241608986

* [x] Where should HTML templates live (`common/templates/discussion` vs. `lms/templates/discussion`)?

  I moved `_discussion_course.html` to `lms/templates/discussion`. It includes some child templates that still live in `common/templates/discussion`; these templates are included by other HTML files that currently live in `common/templates/discussion`.

  Update: As per https://github.com/edx-solutions/edx-platform/pull/741#issuecomment-241608986, moved HTML files that are present on master fork back to original directory, and bundled Mustache template with code that is unique to solutions fork.

* [x] Where should CSS files live, and if they should live in `xblock_discussion/static/css`, which mechanism can/should be used to load them?

  Update: Leaving CSS files where they are for now; moving them is part of YONK-382, cf. https://github.com/edx-solutions/edx-platform/pull/741#issuecomment-241608986

-->

**Console errors**

The following errors come up in the JS console when loading a unit containing an instance of `DiscussionCourseXBlock` in the LMS:

* Unable to load Sequence: `DiscussionApp` is not defined
* backbone.js:1659: `Uncaught Error: Backbone.history has already been started`
* jquery.js:5434: `GET http://localhost:8000/c4x/OpenCraft/IDX101/asset/sass_discussion-forum.css` fails (not sure why it's trying to load this file)

They are present on the parent branch as well, along with some errors about additional discussion-related resources being missing.

There is a dedicated story for addressing these errors: [YONK-401](https://openedx.atlassian.net/browse/YONK-401)

**Discussion tab**

The course-wide discussion tab (the built-in tab, not `DiscussionCourseXBlock`) is currently broken.
This is unrelated to the changes introduced here; the course-wide discussion tab is broken on the parent branch as well (same errors in console, nothing happens when clicking "Add a Post").

bradenmacdonald commented on common/static/common/js/discussion/discussion_router.js:
```diff
@@ -129,9 +128,7 @@
-                return this.navigate("" + (thread.get("commentable_id")) + "/threads/" + thread_id, {
-                    trigger: true
-                });
+                return this.showThread(thread);
```
Why did this have to be changed?

itsjeyd replied:
@bradenmacdonald This had to be changed because for some reason, `showThread` does not have access to an up-to-date thread collection, so it couldn't show a newly added thread.

(To be more specific, `showThread` would try to retrieve the thread via `this.discussion.get(thread_id)`, using `thread_id` from the `:forum_name/threads/:thread_id` URL. However, `this.discussion` did not contain an entry for the newly added thread, so the lookup would fail).

bradenmacdonald replied:
@itsjeyd But the new code is still using `this.discussion.get(thread_id);` - it's just doing that lookup in `navigateToThread` instead of in `showThread`. So I don't see why this would fix it?

bradenmacdonald replied:
Also, it seems like with this change, the URL is not going to get updated when the user clicks on the thread.

itsjeyd replied:
@bradenmacdonald The difference between `navigateToThread` and `showThread

I pushed an update that allows the URL to change when clicking on a specific thread.

itsjeyd replied:
@bradenmacdonald The difference between `navigateToThread` and `showThread` is that `navigateToThread` has access to an up-to-date thread collection, so doing the lookup inside that method will succeed, and `showThread` doesn't. I don't know why `showThread` only has access to stale data, but I think this issue might be related to the `Uncaught Error: Backbone.history has already been started` error that comes up in the console. From the Backbone.js documentation on `Backbone.Router`:

> During page load, after your application has finished creating all of its routers, be sure to call `Backbone.history.start()` or `Backbone.history.start({pushState: true})` to route the initial URL.

`navigateToThread` does not act as a handler for a route (it is not mentioned in `DiscussionRouter.prototype.routes`), so it wouldn't be affected by any problems resulting from the fact that the `Backbone.history.start` call fails with the error mentioned above. So far, this is the only hint I found that might explain why doing the lookup inside `showThread` produces a different result than doing the lookup inside `navigateToThread`.

> Also, it seems like with this change, the URL is not going to get updated when the user clicks on the thread.

I pushed an update that allows the URL to change when clicking on a specific thread: https://github.com/edx-solutions/edx-platform/pull/741/commits/ecb6ce70f662dced72897bba6175a59a6f411fde.

However, since you mentioned [here](https://github.com/edx-solutions/edx-platform/pull/741#issuecomment-242170680) that we shouldn't implement too many fixes until [YONK-372](https://openedx.atlassian.net/browse/YONK-372) is completed, and the existing changes are more of a workaround, I'd also be fine with *not* trying to address the problem of not being able to show a newly created thread right away in the context of this PR. Let me know.

bradenmacdonald replied:
@itsjeyd Yeah, I'd rather understand the issue and do a proper fix in YONK-401 than use a workaround now and forget to fix it properly in the future. And we should be aiming to make this fork more similar to upstream, with a smaller diff.

bradenmacdonald commented:
@itsjeyd 

Can you please summarize the changes you've made in the PR description? Some parts of the diff are hard to understand without an explanation.

> Change package structure to mimic other XBlocks? (Put xblock_discussion.py, xblock_discussion_course.py, etc. into subfolder of xblock_discussion, leave setup.py at top level)

I don't object, but I don't see any benefit to doing that; personally, I prefer to organize into folders only when necessary, rather than making a hierarchy when there's only a couple files. If we do do this, however, we would have to first do it for the upstream edx-platform, then do it for this solutions fork. (Our goal here above all is to reduce the diff between master and solutions).

So: No, not for now.

> Where should HTML templates live?

For mako templates, it doesn't really matter; anything that's only read and rendered on the server-side can be put anywhere. The things to consider are:
* we want the diff with master to be as small as possible, so if the template exists on master, then leave it in that place / move it to the same place.
* If the static asset build system or the includes used in the templates require all templates to be in a specific place, then move/leave them there.
* Otherwise, move them to `xblock_discussion/templates`

For the moustache templates, I think they should be moved to the `xblock_discussion` folder - they're not on the master fork, and moving them to that folder helps keep the solutions fork's code clean. **Actually,** on thinking about this, we may want to aim to keep the `xblock_discussion` folder unmodified (as close to `master` as possible), and put the course-wide xblock in a sibling `xblock_course_discussion` folder - which would contain all the XBlock code and all js/css/html that is unique to the course-wide block on the solutions fork (if possible). What do you think?

> Where should CSS files live, and if they should live in xblock_discussion/static/css, which mechanism can/should be used to load them?

That's part of YONK-382, so you don't have to do it now, but yes, anything which is specific to the solutions fork should be moved into `xblock_discussion/public/css` and delivered via the XBlock Fragment API. (I believe that was already the case for the course-wide discussion xblock? Otherwise it would not have worked in apros.) Any CSS which is shared with the master fork should also be moved to the same place, but we need to do that on the master fork first, then backport it to this fork. (Again, that's part of YONK-382).

itsjeyd commented:
@bradenmacdonald 

> ... anything which is specific to the solutions fork should be ... delivered via the XBlock Fragment API. (I believe that was already the case for the course-wide discussion xblock? ...)

Yes, that's right: The CSS files for the course-wide discussion xblock still live in "global" directories (`lms/static/css` and `common/static/xblock/discussion/css`), but they are added to the fragment via `add_css_url`.

itsjeyd commented:
@bradenmacdonald 

> we may want to aim to keep the xblock_discussion folder unmodified (as close to master as possible), and put the course-wide xblock in a sibling xblock_course_discussion folder

This works (cf. ) and is probably the best solution w/r/t reducing the potential for conflicts between the master and solution forks. The only thing that becomes less straightforward with this solution is sharing CSS resources between in-line and course-wide discussion blocks, but I'm not sure if we'd even want to do that?

itsjeyd commented:
Thanks @bradenmacdonald for the feedback. I addressed your comments and updated the PR description.

bradenmacdonald commented:
> the PR also fixes a few errors related to rendering the block (1) and creating new posts

We probably shouldn't do too much on that until [YONK-372](https://openedx.atlassian.net/browse/YONK-372) is completed.

bradenmacdonald commented:
OK, two last things before +1 from me:

1. The change to `DiscussionRouter` still doesn't look right to me - see comment.
2. The course-wide discussion tab (the built-in tab, not this XBlock) is broken - is that related to these changes or no?

FYI, I created YONK-401 to track and fix those console errors, once the cs_comments_service rebase is complete.

bradenmacdonald commented:
3. What was the rationale of ffc43edde72473cc028f9e8667c8ff1a2febe5f5  ? Can we not just link to the existing file (`lms/static/css/discussion-forum.css`) as you were previously doing? I don't want to duplicate that file, especially since it's generated from SCSS and isn't a source file.

itsjeyd commented:
@bradenmacdonald Thanks for reviewing!

> The change to DiscussionRouter still doesn't look right to me - see comment.

Replied directly to your comment.

> The course-wide discussion tab (the built-in tab, not this XBlock) is broken - is that related to these changes or no?

It's not related to these changes; I double-checked the parent branch and it's broken there as well (same errors in console, nothing happens when clicking "Add a Post"). Added a note about that to the PR description.

> What was the rationale of ffc43edde72473cc028f9e8667c8ff1a2febe5f5  ? Can we not just link to the existing file (`lms/static/css/discussion-forum.css`) as you were previously doing? I don't want to duplicate that file, especially since it's generated from SCSS and isn't a source file.

Linking to the existing file was working for me as well, prior to rebasing this branch to pull in recent changes. Now, doing `git clean -fx` followed by `paver devstack lms` does not add a `discussion-forum.css` file to the `lms/static/css/` folder. I noticed this because all of a sudden there was no border around the area that displays individual posts anymore (same behavior on the parent branch). That's why I added the file to `xblock_discussion_course`.

bradenmacdonald commented:
@itsjeyd I think that's because the SCSS files were moved around; as far as I can tell, `lms/static/css/discussion/lms-discussion-main.css` is the new replacement for `discussion-forum.css`. So please use that in-situ now if it works, and if not, stick with copying as you did in ffc43edde72473cc028f9e8667c8ff1a2febe5f5 and make a note on YONK-401 to remove that copy if possible in favor of the upstream `lms-discussion-main.css`.

bradenmacdonald commented:
:+1:

I notice now that the backbone.history changes the URL when I click on a thread, so that refreshing the page actually leaves the XBlock and goes to the "real" course-wide discussion component. So contrary to my earlier thoughts, we will actually need to disable that functionality in YONK-401. Anyhow, seems like you've achieved the goal here.

- [x] I tested this: Tested in LMS on a Eucalyptus Solutions devstack
- [x] I read through the code
- [x] Includes documentation: n/a

bradenmacdonald commented:
@ziafazal Would you be able to provide a quick second review of this, and give us the OK to merge to your branch?

The goal of this PR was just to move the course-wide discussion XBlock to the same place as the new Discussion XBlock that's in edx-platform. Further bugfixes will be done in YONK-401, once Muhammad has completed YONK-372.

ziafazal commented:
LGTM 👍

itsjeyd commented:
Thanks @ziafazal, merging to your branch now.
